src/connection.rs

```diff
@@ -733,56 +735,73 @@ impl<Exe: Executor> Connection<Exe> {
-            }
+            Some(Some(addresses)) if !addresses.is_empty() => addresses,
+            _ => return Err(ConnectionError::NotFound)
         };
 
-        let hostname = hostname.unwrap_or_else(|| address.ip().to_string());
-
         let id = Uuid::new_v4();
-        debug!("Connecting to {}: {}, as {}", url, address, id);
-        let sender_prepare = Connection::prepare_stream(
-            id,
-            address,
-            hostname,
-            tls,
-            auth_data,
-            proxy_to_broker_url,
-            certificate_chain,
-            allow_insecure_connection,
-            tls_hostname_verification_enabled,
-            executor.clone(),
-            operation_timeout,
-        );
-        let delay_f = executor.delay(connection_timeout);
-
-        pin_mut!(sender_prepare);
-        pin_mut!(delay_f);
-
-        let sender = match select(sender_prepare, delay_f).await {
-            Either::Left((res, _)) => res?,
-            Either::Right(_) => {
-                warn!("TimedOut connecting to the pulsar server {}", line!());
-                return Err(ConnectionError::Io(std::io::Error::new(
-                    std::io::ErrorKind::TimedOut,
-                    "timeout connecting to the Pulsar server",
-                )));
-            }
-        };
+        let mut errors = vec![];
+        for address in addresses {
+            let hostname = hostname.clone().unwrap_or_else(|| address.ip().to_string());
+            debug!("Connecting to {}: {}, as {}", url, address, id);
+            let sender_prepare = Connection::prepare_stream(
+                id,
+                address,
+                hostname,
+                tls,
+                auth_data.clone(),
+                proxy_to_broker_url.clone(),
+                certificate_chain,
+                allow_insecure_connection,
+                tls_hostname_verification_enabled,
+                executor.clone(),
+                operation_timeout,
+            );
+            let delay_f = executor.delay(connection_timeout);
+
+            pin_mut!(sender_prepare);
+            pin_mut!(delay_f);
+
+            let sender;
+            match select(sender_prepare, delay_f).await {
+                Either::Left((Ok(res), _)) => sender = res,
+                Either::Left((Err(err), _)) => {
+                    errors.push(err);
+                    continue;
+                }
+                Either::Right(_) => {
+                    errors.push(ConnectionError::Io(std::io::Error::new(
+                        std::io::ErrorKind::TimedOut,
+                        "timeout connecting to the Pulsar server",
+                    )));
+                    continue;
+                }
+            };
 
-        Ok(Connection { id, url, sender })
+            return Ok(Connection { id, url, sender })
+        }
+
+        let retryable_errors = errors
+            .drain_filter(|e| e.establish_retryable())
```

Comment:
To me, this line should be written like below to remove the nightly requirement:
```suggestion
            .drain()
            .filter(|e| e.establish_retryable())
```

Reply:
Not exactly. `drain` take a range instead of a condition. And if we move the original errors, we cannot use it in the line:

```rust
error!("connection error, not retryable: {:?}", errors);
```

But perhaps we don't have to do error handling so fine-grained.

Reply:
But we can simply write some imperative code here. I'm preparing a commit on this.